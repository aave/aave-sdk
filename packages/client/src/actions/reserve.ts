import { type Reserve, ReserveQuery, type ReserveRequest } from '@aave/graphql';
import type { ResultAsync } from '@aave/types';
import type { AaveClient } from '../client';
import type { UnexpectedError } from '../errors';

/**
 * Fetches a specific reserve by market address, token address, and chain ID.
 *
 * ```ts
 * const result = await reserve(client, {
 *   market: evmAddress('0x87870bca3f3fd6335c3f4ce8392d69350b4fa4e2'),
 *   token: evmAddress('0xa0b86a33e6441c8c5f0bb9b7e5e1f8bbf5b78b5c'),
 *   chainId: chainId(1)
 * });
 * ```
 *
 * @param client - Aave client.
 * @param request - The reserve request parameters.
 * @returns The reserve data, or null if not found.
 */
export function reserve(
  client: AaveClient,
  request: ReserveRequest,
): ResultAsync<Reserve | null, UnexpectedError> {
  return client.query(ReserveQuery, { request });
}
