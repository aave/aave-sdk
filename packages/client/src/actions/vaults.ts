import type { UnexpectedError } from '@aave/core';
import {
  type PaginatedVaultsResult,
  type PaginatedVaultUserTransactionHistoryResult,
  type TokenAmount,
  type TransactionRequest,
  UserVaultsQuery,
  type UserVaultsRequest,
  type Vault,
  type VaultDeployRevenueSplitterRequest,
  VaultPreviewDepositQuery,
  type VaultPreviewDepositRequest,
  VaultPreviewMintQuery,
  type VaultPreviewMintRequest,
  VaultPreviewRedeemQuery,
  type VaultPreviewRedeemRequest,
  VaultPreviewWithdrawQuery,
  type VaultPreviewWithdrawRequest,
  VaultQuery,
  type VaultRequest,
  type VaultSetRevenueSplitterRequest,
  VaultsQuery,
  type VaultsRequest,
  type VaultTransferOwnershipRequest,
  VaultUserActivityQuery,
  type VaultUserActivityRequest,
  type VaultUserActivityResult,
  VaultUserTransactionHistoryQuery,
  type VaultUserTransactionHistoryRequest,
  vaultDeployRevenueSplitterQuery,
  vaultSetRevenueSplitterQuery,
  vaultTransferOwnershipQuery,
} from '@aave/graphql';
import type { ResultAsync } from '@aave/types';
import type { AaveClient } from '../AaveClient';

/**
 * Fetches a specific vault by address and chain ID.
 *
 * **Example: By Address**
 * ```ts
 * const result = await vault(client, {
 *   by: {
 *     address: evmAddress('0x1234…'),
 *   },
 *   chainId: chainId(1),
 *   user: evmAddress('0x5678…'),
 * });
 * ```
 *
 * **Example: Tx Hash**
 * ```ts
 * const result = await vault(client, {
 *   by: {
 *     txHash: txHash('0x1234…'),
 *   },
 *   chainId: chainId(1),
 *   user: evmAddress('0x5678…'),
 * });
 * ```
 *
 * @param client - Aave client.
 * @param request - The vault request parameters.
 * @returns The vault data, or null if not found.
 */
export function vault(
  client: AaveClient,
  request: VaultRequest,
): ResultAsync<Vault | null, UnexpectedError> {
  return client.query(VaultQuery, {
    request,
  });
}

/**
 * Fetches vaults based on filter criteria.
 *
 * ```ts
 * const result = await vaults(client, {
 *   criteria: {
 *     ownedBy: [evmAddress('0x1234…')]
 *   },
 *   pageSize: PageSize.Ten,
 *   user: evmAddress('0x5678…'),
 * });
 * ```
 *
 * @param client - Aave client.
 * @param request - The vaults request parameters.
 * @returns The paginated vaults result.
 */
export function vaults(
  client: AaveClient,
  request: VaultsRequest,
): ResultAsync<PaginatedVaultsResult, UnexpectedError> {
  return client.query(VaultsQuery, {
    request,
  });
}

/**
 * Fetches vaults that a user has shares in.
 *
 * ```ts
 * const result = await userVaults(client, {
 *   user: evmAddress('0x1234…'),
 *   filters: {
 *     markets: [evmAddress('0x5678…')]
 *   },
 *   orderBy: { shares: OrderDirection.Desc },
 *   pageSize: PageSize.Fifty,
 * });
 * ```
 *
 * @param client - Aave client.
 * @param request - The user vaults request parameters.
 * @returns The paginated user vaults result.
 */
export function userVaults(
  client: AaveClient,
  request: UserVaultsRequest,
): ResultAsync<PaginatedVaultsResult, UnexpectedError> {
  return client.query(UserVaultsQuery, {
    request,
  });
}

/**
 * Determines the amount of shares that would be received for a deposit.
 *
 * ```ts
 * const result = await vaultPreviewDeposit(client, {
 *   vault: evmAddress('0x1234567890abcdef1234567890abcdef12345678'),
 *   chainId: chainId(1),
 *   amount: bigDecimal('1000'),
 * });
 *
 * if (result.isOk()) {
 *   console.log('Shares to receive:', result.value.amount.value);
 *   console.log('USD value:', result.value.usd);
 * }
 * ```
 *
 * @param client - Aave client.
 * @param request - The vault preview deposit request parameters.
 * @returns The simulated shares amount that would be received.
 */
export function vaultPreviewDeposit(
  client: AaveClient,
  request: VaultPreviewDepositRequest,
): ResultAsync<TokenAmount, UnexpectedError> {
  return client.query(VaultPreviewDepositQuery, {
    request,
  });
}

/**
 * Determines the amount of assets that would be required to mint a specific amount of vault shares.
 *
 * ```ts
 * const result = await vaultPreviewMint(client, {
 *   vault: evmAddress('0x1234567890abcdef1234567890abcdef12345678'),
 *   chainId: chainId(1),
 *   amount: bigDecimal('500'),
 * });
 *
 * if (result.isOk()) {
 *   console.log('Assets required:', result.value.amount.value);
 *   console.log('USD value:', result.value.usd);
 * }
 * ```
 *
 * @param client - Aave client.
 * @param request - The vault preview mint request parameters.
 * @returns The simulated assets amount that would be required.
 */
export function vaultPreviewMint(
  client: AaveClient,
  request: VaultPreviewMintRequest,
): ResultAsync<TokenAmount, UnexpectedError> {
  return client.query(VaultPreviewMintQuery, {
    request,
  });
}

/**
 * Determines the amount of shares that would be burned for a withdrawal.
 *
 * ```ts
 * const result = await vaultPreviewWithdraw(client, {
 *   vault: evmAddress('0x1234567890abcdef1234567890abcdef12345678'),
 *   chainId: chainId(1),
 *   amount: bigDecimal('750'),
 * });
 *
 * if (result.isOk()) {
 *   console.log('Shares to burn:', result.value.amount.value);
 *   console.log('USD value:', result.value.usd);
 * }
 * ```
 *
 * @param client - Aave client.
 * @param request - The vault preview withdraw request parameters.
 * @returns The simulated shares amount that would be burned.
 */
export function vaultPreviewWithdraw(
  client: AaveClient,
  request: VaultPreviewWithdrawRequest,
): ResultAsync<TokenAmount, UnexpectedError> {
  return client.query(VaultPreviewWithdrawQuery, {
    request,
  });
}

/**
 * Determines the amount of assets that would be received for redeeming a specific amount of vault shares.
 *
 * ```ts
 * const result = await vaultPreviewRedeem(client, {
 *   vault: evmAddress('0x1234567890abcdef1234567890abcdef12345678'),
 *   chainId: chainId(1),
 *   amount: bigDecimal('200'),
 * });
 *
 * if (result.isOk()) {
 *   console.log('Assets to receive:', result.value.amount.value);
 *   console.log('USD value:', result.value.usd);
 * }
 * ```
 *
 * @param client - Aave client.
 * @param request - The vault preview redeem request parameters.
 * @returns The simulated assets amount that would be received.
 */
export function vaultPreviewRedeem(
  client: AaveClient,
  request: VaultPreviewRedeemRequest,
): ResultAsync<TokenAmount, UnexpectedError> {
  return client.query(VaultPreviewRedeemQuery, {
    request,
  });
}

/**
 * Fetches the user transaction history for a vault.
 *
 * ```ts
 * const result = await vaultUserTransactionHistory(client, {
 *   vault: evmAddress('0x1234567890abcdef1234567890abcdef12345678'),
 *   chainId: chainId(1),
 *   user: evmAddress('0x5678901234567890abcdef1234567890abcdef12'),
 * });
 *
 * if (result.isOk()) {
 *   console.log('Transaction count:', result.value.items.length);
 *   result.value.items.forEach(tx => {
 *     if (tx.__typename === 'VaultUserDepositItem') {
 *       console.log('Deposit:', tx.asset.amount.value, 'shares:', tx.shares.amount.value);
 *     } else if (tx.__typename === 'VaultUserWithdrawItem') {
 *       console.log('Withdraw:', tx.asset.amount.value, 'shares:', tx.shares.amount.value);
 *     }
 *   });
 * }
 * ```
 *
 * @param client - Aave client.
 * @param request - The vault user transaction history request parameters.
 * @returns The paginated vault user transaction history result.
 */
export function vaultUserTransactionHistory(
  client: AaveClient,
  request: VaultUserTransactionHistoryRequest,
): ResultAsync<PaginatedVaultUserTransactionHistoryResult, UnexpectedError> {
  return client.query(VaultUserTransactionHistoryQuery, {
    request,
  });
}

/**
 * Fetches the user activity data for a vault, including earnings breakdown over time.
 *
 * ```ts
 * const result = await vaultUserActivity(client, {
 *   vault: evmAddress('0x1234567890abcdef1234567890abcdef12345678'),
 *   chainId: chainId(1),
 *   user: evmAddress('0x5678901234567890abcdef1234567890abcdef12'),
 * });
 *
 * if (result.isOk()) {
 *   console.log('Total earned:', result.value.earned.amount.value);
 *   console.log('Activity breakdown count:', result.value.breakdown.length);
 *   result.value.breakdown.forEach(activity => {
 *     console.log('Date:', activity.date);
 *     console.log('Balance:', activity.balance.amount.value);
 *     console.log('Earned:', activity.earned.amount.value);
 *   });
 * }
 * ```
 *
 * @param client - Aave client.
 * @param request - The vault user activity request parameters.
 * @returns The vault user activity result.
 */
export function vaultUserActivity(
  client: AaveClient,
  request: VaultUserActivityRequest,
): ResultAsync<VaultUserActivityResult, UnexpectedError> {
  return client.query(VaultUserActivityQuery, {
    request,
  });
}

/**
 * Creates a transaction to transfer ownership of a vault.
 *
 * ```ts
 * const result = await vaultTransferOwnership(client, {
 *   vault: evmAddress('0x1234567890abcdef1234567890abcdef12345678'),
 *   chainId: chainId(1),
 *   newOwner: evmAddress('0x5678901234567890abcdef1234567890abcdef12'),
 * }).andThen(sendWith(wallet)).andThen(client.waitForTransaction);
 * ```
 *
 * @param client - Aave client.
 * @param request - The vault transfer ownership request parameters.
 * @returns The transaction data for transferring vault ownership.
 */
export function vaultTransferOwnership(
  client: AaveClient,
  request: VaultTransferOwnershipRequest,
): ResultAsync<TransactionRequest, UnexpectedError> {
  return client.query(vaultTransferOwnershipQuery, {
    request,
  });
}

/**
 * Creates a transaction to deploy a revenue splitter for a vault.
 *
 * ```ts
 * const result = await vaultDeployRevenueSplitter(client, {
 *   vault: evmAddress('0x1234567890abcdef1234567890abcdef12345678'),
 *   chainId: chainId(1),
 *   recipients: [
 *     {
 *       account: evmAddress('0x5678901234567890abcdef1234567890abcdef12'),
 *       shares: 5000, // 50%
 *     },
 *     {
 *       account: evmAddress('0x9abc123456789012345678901234567890abcdef'),
 *       shares: 5000, // 50%
 *     },
 *   ],
 * }).andThen(sendWith(wallet)).andThen(client.waitForTransaction);
 * ```
 *
 * @param client - Aave client.
 * @param request - The vault deploy revenue splitter request parameters.
 * @returns The transaction data for deploying a revenue splitter.
 */
export function vaultDeployRevenueSplitter(
  client: AaveClient,
  request: VaultDeployRevenueSplitterRequest,
): ResultAsync<TransactionRequest, UnexpectedError> {
  return client.query(vaultDeployRevenueSplitterQuery, {
    request,
  });
}

/**
 * Creates a transaction to set the revenue splitter for a vault.
 *
 * ```ts
 * const result = await vaultSetRevenueSplitter(client, {
 *   vault: evmAddress('0x1234567890abcdef1234567890abcdef12345678'),
 *   chainId: chainId(1),
 *   revenueSplitter: evmAddress('0x5678901234567890abcdef1234567890abcdef12'),
 * }).andThen(sendWith(wallet)).andThen(client.waitForTransaction);
 * ```
 *
 * @param client - Aave client.
 * @param request - The vault set revenue splitter request parameters.
 * @returns The transaction data for setting a revenue splitter.
 */
export function vaultSetRevenueSplitter(
  client: AaveClient,
  request: VaultSetRevenueSplitterRequest,
): ResultAsync<TransactionRequest, UnexpectedError> {
  return client.query(vaultSetRevenueSplitterQuery, {
    request,
  });
}
