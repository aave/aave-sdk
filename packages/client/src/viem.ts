import type {
  ExecutionPlan,
  InsufficientBalanceError,
  TransactionRequest,
} from '@aave/graphql';
import {
  errAsync,
  invariant,
  ResultAsync,
  type TxHash,
  txHash,
} from '@aave/types';
import type { Hash, WalletClient } from 'viem';
import {
  sendTransaction as sendEip1559Transaction,
  waitForTransactionReceipt,
} from 'viem/actions';
import { SigningError, ValidationError } from './errors';
import type { ExecutionPlanHandler } from './types';

async function sendTransaction(
  walletClient: WalletClient,
  request: TransactionRequest,
): Promise<Hash> {
  return sendEip1559Transaction(walletClient, {
    account: request.from,
    data: request.data,
    to: request.to,
    value: BigInt(request.value),
    chain: walletClient.chain,
  });
}

/**
 * @internal
 */
export function sendTransactionAndWait(
  walletClient: WalletClient,
  request: TransactionRequest,
): ResultAsync<TxHash, SigningError> {
  // TODO: verify it's on the correct chain, ask to switch if possible
  // TODO: verify if wallet account is correct, switch if possible

  return ResultAsync.fromPromise(
    sendTransaction(walletClient, request),
    (err) => SigningError.from(err),
  )
    .map(async (hash) =>
      waitForTransactionReceipt(walletClient, {
        hash,
        pollingInterval: 100,
        retryCount: 20,
        retryDelay: 50,
      }),
    )
    .map((receipt) => txHash(receipt.transactionHash));
}

/**
 * Sends transactions using the provided wallet client.
 *
 * Handles {@link TransactionRequest} by signing and sending, {@link ApprovalRequired} by sending both approval and original transactions, and returns validation errors for {@link InsufficientBalanceError}.
 */
export function sendWith(
  walletClient: WalletClient | undefined,
): ExecutionPlanHandler {
  return (
    result: ExecutionPlan,
  ): ResultAsync<
    TxHash,
    SigningError | ValidationError<InsufficientBalanceError>
  > => {
    invariant(
      walletClient,
      'Expected a WalletClient to handle the operation result.',
    );

    switch (result.__typename) {
      case 'TransactionRequest':
        return sendTransactionAndWait(walletClient, result);

      case 'ApprovalRequired':
        return sendTransactionAndWait(walletClient, result.approval).andThen(
          () =>
            sendTransactionAndWait(walletClient, result.originalTransaction),
        );

      case 'InsufficientBalanceError':
        return errAsync(ValidationError.fromGqlNode(result));
    }
  };
}
