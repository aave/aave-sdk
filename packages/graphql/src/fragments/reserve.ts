import type { FragmentOf } from 'gql.tada';
import { graphql } from '../graphql';
import { ChainFragment } from './chain';
import {
  CurrencyFragment,
  DecimalValueFragment,
  NativeCurrencyFragment,
  TokenAmountFragment,
} from './common';

export const EmodeInfoFragment = graphql(
  `fragment EmodeInfo on EmodeInfo {
    __typename
    maxLTV {
      ...DecimalValue
    }
    liquidationThreshold {
      ...DecimalValue
    }
    liquidationPenalty {
      ...DecimalValue
    }
  }`,
  [DecimalValueFragment],
);
export type EmodeInfo = FragmentOf<typeof EmodeInfoFragment>;

export const ReserveSupplyInfoFragment = graphql(
  `fragment ReserveSupplyInfo on ReserveSupplyInfo {
    __typename
    apy {
      ...DecimalValue
    }
    total {
      ...DecimalValue
    }
    maxLTV {
      ...DecimalValue
    }
    liquidationThreshold {
      ...DecimalValue
    }
    liquidationPenalty {
      ...DecimalValue
    }
    canBeCollateral
  }`,
  [DecimalValueFragment],
);
export type ReserveSupplyInfo = FragmentOf<typeof ReserveSupplyInfoFragment>;

export const ReserveBorrowInfoFragment = graphql(
  `fragment ReserveBorrowInfo on ReserveBorrowInfo {
    __typename
    apy {
      ...DecimalValue
    }
    total {
      ...DecimalValue
    }
    borrowCap {
      ...DecimalValue
    }
    reserveFactor
  }`,
  [DecimalValueFragment],
);
export type ReserveBorrowInfo = FragmentOf<typeof ReserveBorrowInfoFragment>;

export const ReserveUserAvailabilityFragment = graphql(
  `fragment ReserveUserAvailability on ReserveUserAvailability {
    __typename
    balance {
      ...DecimalValue
    }
    supply {
      ...DecimalValue
    }
    borrow {
      ...DecimalValue
    }
  }`,
  [DecimalValueFragment],
);
export type ReserveUserAvailability = FragmentOf<
  typeof ReserveUserAvailabilityFragment
>;

export const ReserveFragment = graphql(
  `fragment Reserve on Reserve {
    __typename
    market {
      name
      chain {
        ...Chain
      }
    }
    address
    underlyingToken {
      ...Currency
    }
    aToken {
      ...Currency
    }
    vToken {
      ...Currency
    }
    acceptsNative {
      ...NativeCurrency
    }
    size {
      ...TokenAmount
    }
    utilizationRate {
      ...DecimalValue
    }
    available {
      ...TokenAmount
    }
    usdExchangeRate
    usdOracleAddress
    supplyInfo {
      ...ReserveSupplyInfo
    }
    borrowInfo {
      ...ReserveBorrowInfo
    }
    eModeInfo {
      ...EmodeInfo
    }
    userAvailability(address: $userAddress) @include(if: $includeUserFields) {
      ...ReserveUserAvailability
    }
  }`,
  [
    ChainFragment,
    CurrencyFragment,
    NativeCurrencyFragment,
    TokenAmountFragment,
    DecimalValueFragment,
    ReserveSupplyInfoFragment,
    ReserveBorrowInfoFragment,
    EmodeInfoFragment,
    ReserveUserAvailabilityFragment,
  ],
);
export type Reserve = FragmentOf<typeof ReserveFragment>;
