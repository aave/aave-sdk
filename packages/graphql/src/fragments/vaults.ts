import type { FragmentOf } from 'gql.tada';
import { graphql } from '../graphql';
import {
  PaginatedResultInfoFragment,
  PercentValueFragment,
  TokenAmountFragment,
} from './common';
import { ReserveFragment } from './reserve';

export const UserVaultSharesFragment = graphql(
  `fragment UserVaultShares on UserVaultShares {
    __typename
    shares {
      ...TokenAmount
    }
    balance {
      ...TokenAmount
    }
  }`,
  [TokenAmountFragment],
);
export type UserVaultShares = FragmentOf<typeof UserVaultSharesFragment>;

export const VaultFragment = graphql(
  `fragment Vault on Vault {
    __typename
    address
    owner
    shareName
    shareSymbol
    usedReserve {
      ...Reserve
    }
    fee {
      ...PercentValue
    }
    totalFeeRevenue {
      ...TokenAmount
    }
    balance {
      ...TokenAmount
    }
    feesBalance {
      ...TokenAmount
    }
    chainId
    userShares {
      ...UserVaultShares
    }
  }`,
  [
    ReserveFragment,
    PercentValueFragment,
    TokenAmountFragment,
    UserVaultSharesFragment,
  ],
);
export type Vault = FragmentOf<typeof VaultFragment>;

/**
 * @internal
 */
export const PaginatedVaultsResultFragment = graphql(
  `fragment PaginatedVaultsResult on PaginatedVaultsResult {
    __typename
    items {
      ...Vault
    }
    pageInfo {
      ...PaginatedResultInfo
    }
  }`,
  [VaultFragment, PaginatedResultInfoFragment],
);
export type PaginatedVaultsResult = FragmentOf<
  typeof PaginatedVaultsResultFragment
>;
