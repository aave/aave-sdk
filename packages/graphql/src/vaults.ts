import { TokenAmountFragment } from './fragments/common';
import {
  PaginatedVaultsResultFragment,
  type PaginatedVaultUserTransactionHistoryResult,
  PaginatedVaultUserTransactionHistoryResultFragment,
  VaultFragment,
} from './fragments/vaults';
import { graphql, type RequestOf } from './graphql';

/**
 * @internal
 */
export const VaultQuery = graphql(
  `query Vault($request: VaultRequest!) {
    value: vault(request: $request) {
      ...Vault
    }
  }`,
  [VaultFragment],
);
export type VaultRequest = RequestOf<typeof VaultQuery>;

/**
 * @internal
 */
export const VaultsQuery = graphql(
  `query Vaults($request: VaultsRequest!) {
    value: vaults(request: $request) {
      ...PaginatedVaultsResult
    }
  }`,
  [PaginatedVaultsResultFragment],
);
export type VaultsRequest = RequestOf<typeof VaultsQuery>;

/**
 * @internal
 */
export const UserVaultsQuery = graphql(
  `query UserVaults($request: UserVaultsRequest!) {
    value: userVaults(request: $request) {
      ...PaginatedVaultsResult
    }
  }`,
  [PaginatedVaultsResultFragment],
);
export type UserVaultsRequest = RequestOf<typeof UserVaultsQuery>;

/**
 * @internal
 */
export const VaultPreviewDepositQuery = graphql(
  `query VaultPreviewDeposit($request: VaultOperationPreviewRequest!) {
    value: vaultPreviewDeposit(request: $request) {
      ...TokenAmount
    }
  }`,
  [TokenAmountFragment],
);
export type VaultPreviewDepositRequest = RequestOf<
  typeof VaultPreviewDepositQuery
>;

/**
 * @internal
 */
export const VaultPreviewMintQuery = graphql(
  `query VaultPreviewMint($request: VaultOperationPreviewRequest!) {
    value: vaultPreviewMint(request: $request) {
      ...TokenAmount
    }
  }`,
  [TokenAmountFragment],
);
export type VaultPreviewMintRequest = RequestOf<typeof VaultPreviewMintQuery>;

/**
 * @internal
 */
export const VaultPreviewWithdrawQuery = graphql(
  `query VaultPreviewWithdraw($request: VaultOperationPreviewRequest!) {
    value: vaultPreviewWithdraw(request: $request) {
      ...TokenAmount
    }
  }`,
  [TokenAmountFragment],
);
export type VaultPreviewWithdrawRequest = RequestOf<
  typeof VaultPreviewWithdrawQuery
>;

/**
 * @internal
 */
export const VaultPreviewRedeemQuery = graphql(
  `query VaultPreviewRedeem($request: VaultOperationPreviewRequest!) {
    value: vaultPreviewRedeem(request: $request) {
      ...TokenAmount
    }
  }`,
  [TokenAmountFragment],
);
export type VaultPreviewRedeemRequest = RequestOf<
  typeof VaultPreviewRedeemQuery
>;

/**
 * @internal
 */
export const VaultUserTransactionHistoryQuery = graphql(
  `query VaultUserTransactionHistory($request: VaultUserTransactionHistoryRequest!) {
    value: vaultUserTransactionHistory(request: $request) {
      ...PaginatedVaultUserTransactionHistoryResult
    }
  }`,
  [PaginatedVaultUserTransactionHistoryResultFragment],
);
export type VaultUserTransactionHistoryRequest = RequestOf<
  typeof VaultUserTransactionHistoryQuery
>;
export type VaultUserTransactionHistoryResult = {
  value: PaginatedVaultUserTransactionHistoryResult;
};
