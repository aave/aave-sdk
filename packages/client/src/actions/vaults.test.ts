import type { Vault } from '@aave/graphql';
import {
  assertOk,
  bigDecimal,
  evmAddress,
  nonNullable,
  okAsync,
  type ResultAsync,
} from '@aave/types';
import { describe, expect, it } from 'vitest';
import {
  client,
  createNewWallet,
  ETHEREUM_FORK_ID,
  ETHEREUM_MARKET_ADDRESS,
  fundErc20Address,
  getBalance,
  WETH_ADDRESS,
  wait,
} from '../test-utils';
import { sendWith } from '../viem';
import { reserve } from './reserve';
import {
  vaultDeploy,
  vaultDeposit,
  vaultMintShares,
  vaultRedeemShares,
  vaultWithdraw,
} from './transactions';
import { userVaults, vault, vaults } from './vaults';

const organization = createNewWallet();
const user = createNewWallet();

function createVault(): ResultAsync<Vault, Error> {
  return fundErc20Address(
    WETH_ADDRESS,
    evmAddress(organization.account!.address),
    bigDecimal('2'),
  ).andThen(() => {
    return reserve(client, {
      chainId: ETHEREUM_FORK_ID,
      underlyingToken: WETH_ADDRESS,
      market: ETHEREUM_MARKET_ADDRESS,
    }).andThen((reserve) => {
      return vaultDeploy(client, {
        chainId: reserve!.market.chain.chainId,
        market: reserve!.market.address,
        deployer: evmAddress(organization.account!.address),
        owner: evmAddress(organization.account!.address),
        initialFee: bigDecimal('3'),
        initialLockDeposit: bigDecimal('1'),
        shareName: 'Aave WETH Vault Shares',
        shareSymbol: 'avWETH',
        underlyingToken: reserve!.underlyingToken.address,
      })
        .andThen(sendWith(organization))
        .andThen((txHash) =>
          vault(client, { by: { txHash }, chainId: ETHEREUM_FORK_ID }),
        )
        .andTee((vault) => console.log(`vault address: ${vault?.address}`))
        .map(nonNullable);
    });
  });
}

function deposit(amount: number) {
  return (vault: Vault): ResultAsync<Vault, Error> => {
    return fundErc20Address(
      WETH_ADDRESS,
      evmAddress(user.account!.address),
      bigDecimal(amount + 0.1),
    ).andThen(() => {
      return vaultDeposit(client, {
        amount: {
          value: bigDecimal('1'),
          currency: WETH_ADDRESS,
        },
        vault: vault.address,
        depositor: evmAddress(user.account!.address),
        chainId: vault.chainId,
      })
        .andThen(sendWith(user))
        .andTee((txHash) => console.log(`tx to deposit in vault: ${txHash}`))
        .andThen(() => okAsync(vault));
    });
  };
}

function mintShares(amount: number) {
  return (vault: Vault): ResultAsync<Vault, Error> => {
    return fundErc20Address(
      WETH_ADDRESS,
      evmAddress(user.account!.address),
      bigDecimal(amount + 0.1),
    ).andThen(() => {
      return vaultMintShares(client, {
        shares: {
          amount: bigDecimal(amount),
        },
        vault: vault.address,
        minter: evmAddress(user.account!.address),
        chainId: vault.chainId,
      })
        .andThen(sendWith(user))
        .andTee((tx) => console.log(`tx to mint shares: ${tx}`))
        .andThen(() => okAsync(vault));
    });
  };
}

describe('Given the Aave Vaults', () => {
  describe('When an organization deploys a new vault', () => {
    it('Then it should be available in the organization vaults', async () => {
      const initialVault = await createVault();
      assertOk(initialVault);

      const result = await vaults(client, {
        criteria: {
          ownedBy: [evmAddress(organization.account!.address)],
        },
      });

      assertOk(result);
      expect(result.value.items[0]?.owner).toEqual(
        organization.account!.address,
      );
      expect(result.value.items[0]?.address).toEqual(
        initialVault.value?.address,
      );
    });
  });

  describe('And a deployed organization vault', () => {
    describe('When a user deposits into the vault', () => {
      it(`Then the operation should be reflected in the user's vault positions`, async () => {
        const initialVault = await createVault()
          .andThen(deposit(1))
          .andTee(() => wait(2000)); // wait for the deposit to be processed
        assertOk(initialVault);

        const userPositions = await userVaults(client, {
          user: evmAddress(user.account!.address),
        });

        assertOk(userPositions);
        expect(userPositions.value.items.length).toBe(1);
        expect(
          userPositions.value.items[0]?.balance.amount.value,
        ).toBeBigDecimalCloseTo(1, 2);
      });
    });

    describe(`When the user mints some vault's shares`, () => {
      it(`Then the operation should be reflected in the user's vault positions`, async () => {
        const initialVault = await createVault()
          .andThen(mintShares(1))
          .andTee(() => wait(2000)); // wait for the mint to be processed
        assertOk(initialVault);

        const userPositions = await userVaults(client, {
          user: evmAddress(user.account!.address),
        });
        assertOk(userPositions);
        expect(
          userPositions.value.items[0]?.userShares?.shares.amount.value,
        ).toBeBigDecimalCloseTo(1, 4);
      });
    });

    describe('When the user withdraws their assets from the vault', () => {
      it.skip(`Then the operation should be reflected in the user's vault positions`, async () => {
        const initialVault = await createVault()
          .andThen(deposit(1))
          .andTee(() => wait(2000)); // wait for the deposit to be processed
        assertOk(initialVault);

        const balanceBefore = await getBalance(
          evmAddress(user.account!.address),
          WETH_ADDRESS,
        );
        const withdrawResult = await vaultWithdraw(client, {
          chainId: initialVault.value?.chainId,
          sharesOwner: evmAddress(user.account!.address),
          underlyingToken: {
            asAToken: false,
            amount: bigDecimal('0.5'),
          },
          vault: initialVault.value?.address,
        })
          .andTee((result) =>
            console.log(`result: ${JSON.stringify(result, null, 2)}`),
          )
          .andThen(sendWith(user))
          .andTee((tx) => console.log(`tx to withdraw from vault: ${tx}`))
          .andTee(() => wait(2000)); // wait for the withdraw to be processed
        assertOk(withdrawResult);

        const userPositions = await userVaults(client, {
          user: evmAddress(user.account!.address),
        });
        assertOk(userPositions);
        const balanceAfter = await getBalance(
          evmAddress(user.account!.address),
          WETH_ADDRESS,
        );
        expect(balanceAfter).toBeGreaterThan(balanceBefore);
        expect(
          userPositions.value.items[0]?.balance.amount.value,
        ).toBeBigDecimalCloseTo(0, 2);

        // TODO: check balance in the wallet after withdraw
      });
    });

    describe('When the user redeems their shares', () => {
      it.skip(`Then the operation should be reflected in the user's vault positions`, async () => {
        const initialVault = await createVault()
          .andThen(mintShares(1))
          .andTee(() => wait(2000)); // wait for the mint to be processed
        assertOk(initialVault);

        const redeemResult = await vaultRedeemShares(client, {
          shares: {
            amount: bigDecimal('1'),
          },
          vault: initialVault.value!.address,
          chainId: initialVault.value!.chainId,
          sharesOwner: evmAddress(user.account!.address),
        })
          .andThen(sendWith(user))
          .andTee((tx) => console.log(`tx to redeem shares: ${tx}`))
          .andTee(() => wait(2000)); // wait for the redeem to be processed
        assertOk(redeemResult);

        const userPositions = await userVaults(client, {
          user: evmAddress(user.account!.address),
        });
        assertOk(userPositions);
        expect(
          userPositions.value.items[0]?.userShares?.shares.amount.value,
        ).toBeBigDecimalCloseTo(0, 4);
      });
    });

    describe(`When the organization changes the vault's fee`, () => {
      it.todo(
        'Then the new fee should be reflected in the vault object',
        async () => {
          // assert vault.fee
        },
      );
    });

    describe('When users borrow from the underlying vault reserve', () => {
      // const borrower = createNewWallet();

      it.todo('Then the vault should accrue its fees', () => {
        // const setup = createVault().andThen(deposit(100));
        // assertOk(result);
        // assert vault.totalFeeRevenue
      });
    });

    describe(`When the organization withdraws the vault's fees`, () => {
      // const borrower = createNewWallet();

      it.todo(
        'Then they shoudl receive the expected ERC-20 amount',
        async () => {},
      );
    });
  });
});
