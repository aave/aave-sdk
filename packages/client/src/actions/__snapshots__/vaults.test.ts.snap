// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given the Aave Vaults > And a deployed organization vault > When a user previews a deposit into the vault > Then it should return the expected amount of shares 1`] = `
{
  "__typename": "TokenAmount",
  "amount": Any<Object>,
  "usd": Any<String>,
  "usdPerToken": Any<String>,
}
`;

exports[`Given the Aave Vaults > And a deployed organization vault > When a user previews a minting shares from a vault > Then it should return the expected amount of tokens needed to mint 1`] = `
{
  "__typename": "TokenAmount",
  "amount": Any<Object>,
  "usd": Any<String>,
  "usdPerToken": Any<String>,
}
`;

exports[`Given the Aave Vaults > And a deployed organization vault > When a user previews a redeeming shares from a vault > Then it should return the expected amount of assets to receive 1`] = `
{
  "__typename": "TokenAmount",
  "amount": Any<Object>,
  "usd": Any<String>,
  "usdPerToken": Any<String>,
}
`;

exports[`Given the Aave Vaults > And a deployed organization vault > When a user previews a withdrawal assets from a vault > Then it should return the expected amount of shares to burn 1`] = `
{
  "__typename": "TokenAmount",
  "amount": Any<Object>,
  "usd": Any<String>,
  "usdPerToken": Any<String>,
}
`;
